File: controllers/ilqr/scripts/module/ilqr.py
#!/usr/bin/env python3

import rospy
import time

from nav_msgs.msg import Odometry, Path
from geometry_msgs.msg import TwistStamped, PoseStamped

from tf.transformations import euler_from_quaternion, quaternion_from_euler

from module.cost import *
from module.dynamics import *

np.set_printoptions(suppress=True)


class iLQR():
    def __init__(self, dynamics, cost, dt=0.1, N=10, v_max=1.0, omega_max=1.0, regu_init=20, regu_max=1000):
        """Constructs an iLQR controller.

        Args:
            dynamics: Dynamics System.
            cost: Cost function.
            dt: Sampling time
            N: Horizon length.
        """

        self.dynamics = dynamics
        self.cost = cost

        self.N = N
        self.alpha = 1.0
        self.regu_init = regu_init
        self.regu_max = regu_max

        self.v_max = v_max
        self.omega_max = omega_max

        self._k = np.zeros((N, 3))
        self._K = np.zeros((N, 2, 3))

        # Publishers and Subscribers
        self.odom_topic = rospy.get_param("odom/frame_id")
        self.odom_subscriber = rospy.Subscriber(self.odom_topic, Odometry, self.odom_callback)
        self.goal_subscriber = rospy.Subscriber("/terrasentia/goal", PoseStamped, self.goal_callback)

        self.path_publisher     = rospy.Publisher("/terrasentia/path", Path, queue_size=1)
        self.cmd_vel_publisher  = rospy.Publisher("/terrasentia/cmd_vel", TwistStamped, queue_size=10)

        self.odom    = Odometry()
        self.goal    = PoseStamped()
        self.cmd_vel = TwistStamped()

    def run(self):
        start_time = time.time()

        x_0 = self.state_from_pose(self.odom.pose.pose)
        xref = self.state_from_pose(self.goal.pose)

        x0 = x_0 - xref

        us = np.zeros((self.N, 2))

        x, u = self.fit(x0, us)

        x += xref

        path_msg = Path()
        path_msg.header.stamp = rospy.Time.now()
        path_msg.header.frame_id = self.odom_topic
        for xi in x:
            pose = self.pose_from_state(xi)
            path_msg.poses.append(pose)

        self.cmd_vel.twist.linear.x = u[0][0]
        self.cmd_vel.twist.angular.z = u[0][1]

        self.path_publisher.publish(path_msg)
        self.cmd_vel_publisher.publish(self.cmd_vel)

        end_time = time.time()
        elapsed_time = end_time - start_time
        print(f'iLQR Run | State {x0} | Reference {xref} | Action control {u[0]} | Time {elapsed_time:.3f} seconds')

    def fit(self, x0, us, n_iter=10, tol=1e-6):
        """Compute optimal control

        Args:
            x0: Initial state.
            u0: Initial control path.
            n_iter: Max iterations.
            tol: Tolerance. 

        Returns:
            xs: Optimal state path.
            us: Optimal control path.
        """
        alphas = 1.1**(-np.arange(10)**2)

        regu = self.regu_init

        # Rollout for first J
        xs, J_old = self._rollout(x0, us)

        # Action control compute loop
        for iteration in range(n_iter):
            k, K, _ = self._backward_pass(xs, us, regu)

            for alpha in alphas:
                xs_new, us_new = self._forward_pass(xs, us, k, K, alpha)
                J = self.cost._trajectory_cost(xs_new, us_new)

                # print(f'Trying convergence {J} on {J_old}')
                
                if (abs(J) < abs(J_old)):

                    J_old = J
                    xs = xs_new
                    us = us_new
                    regu *= 0.7

                    break
            
            # Increase regularization
            else:
                regu *= 2.0

        return xs, us

    def _forward_pass(self, xs, us, ks, Ks, alpha, hessian=False):
        """Apply the forward dynamics

        Args:
            xs: Initial state.
            us: Control path.
            ks: Feedforward gains.
            Ks: Feedback gains.
            alpha: Line search coefficient.
            
        Returns:
            Tuple of:
                x_new: New state path.
                u_new: New action control.
        """
        
        u_new = np.empty((self.N, 2))
        x_new = np.empty((self.N+1, 3))

        x_new[0] = np.copy(xs[0])

        # Compute new action and state control
        for i in range (self.N):

            #print(alpha * ks[i])
            u_new[i] = us[i] + Ks[i] @ (x_new[i] - xs[i]) + alpha * ks[i]

            if (u_new[i][0] < 0):
                u_new[i][0] = 0
            elif (u_new[i][0] > self.v_max):
                u_new[i][0] = self.v_max

            if (u_new[i][1] < -self.omega_max):
                u_new[i][1] = -self.omega_max
            elif (u_new[i][1] > self.omega_max):
                u_new[i][1] = self.omega_max

            x_new[i+1] = np.array(self.dynamics.f(x_new[i], u_new[i])).T

        # print(u_new)
        # print(x_new)

        return x_new, u_new



    def _backward_pass(self, xs, us, regu):
        """Computes the feedforward and feedback gains k and K.

        Args:
            xs: Initial state.
            us: Control path.

        Returns:
            ks: feedforward gains.
            Ks: feedback gains.
            J: Cost path.
        """
        
        N = self.N

        ks = np.empty(us.shape)
        Ks = np.empty((us.shape[0], us.shape[1], xs.shape[1]))

        J = 0

        Qf = self.cost.get_Qf()

        s = Qf @ xs[N-1]
        S = Qf

        regu_I = regu * np.eye(2)

        for n in range(N - 1, -1, -1):

            # Obtain f and l derivatives
            Ak, Bk = self.dynamics.f_prime(xs[n], us[n])
            l_x, l_u, l_xx, l_ux, l_uu  = self.cost.l_prime(xs[n], us[n])

            # Q_terms
            Q_x  = l_x  + np.dot(s, Ak)
            Q_u  = l_u  + np.dot(s, Bk)
            Q_xx = l_xx + np.dot(Ak.T, np.dot(S, Ak))
            Q_uu = l_uu + np.dot(Bk.T, np.dot(S, Bk))
            Q_ux = l_ux + np.dot(Bk.T, np.dot(S, Ak))

            # Regularization
            Q_uu_reg = Q_uu + regu_I       

            # Feedforward and feedback gains
            k = np.dot(-np.linalg.inv(Q_uu), Q_u)
            K = np.dot(-np.linalg.inv(Q_uu), Q_ux)

            ks[n], Ks[n] = k, K

            # V_terms
            s = Q_x + K.T @ Q_u + Q_ux.T @ k + K.T @ Q_uu @ k
            S = Q_xx + K.T@Q_ux + K.T @ Q_ux + K.T @ Q_uu @ K

            #Sum cost 
            J += 1/2 * np.dot(k.T, np.dot(Q_uu, k)) + np.dot(k, Q_u)

        return ks, Ks, J
        
    def _rollout(self, x0, us):
        """Apply the forward dynamics to have a trajectory from the starting
        state x0 by applying the control path us.

        Args:
            x0: Initial state.
            us: Control path.

        Returns:
            J: Cost path.
        """
        J = 0
        N = self.N

        xs = np.empty((N+1, 3))
        xs[0] = x0

        # Calculate path state over a x0 and us
        for n in range(N):
            xs[n+1] = np.array(self.dynamics.f(xs[n], us[n])).T

            J += self.cost._l(xs[n], us[n])[0]
            # print(f'Index {n} | cost: {J}')

        J += self.cost._lf(xs[N-1])
        # print(f'Index {N-1} | cost: {J}')

        return xs, J

    def state_from_pose(self, pose):
        position      = pose.position
        orientation_q = pose.orientation
        orientation = [orientation_q.x, orientation_q.y, orientation_q.z, orientation_q.w]

        x_0 = position.x
        y_0 = position.y
        _, _, theta_0 = euler_from_quaternion(orientation)

        return np.array([x_0, y_0, theta_0])

    def pose_from_state(self, x):
        pose_msg = PoseStamped()
        pose_msg.header.stamp = rospy.Time.now()
        pose_msg.header.frame_id = self.odom_topic
   
        pose_msg.pose.position.x = x[0]
        pose_msg.pose.position.y = x[1] 

        orientation = quaternion_from_euler(0.0, 0.0, x[2])
        pose_msg.pose.orientation.x = orientation[0]
        pose_msg.pose.orientation.y = orientation[1]
        pose_msg.pose.orientation.z = orientation[2]
        pose_msg.pose.orientation.w = orientation[3]

        return pose_msg
    
    def odom_callback(self, msg):
        self.odom = msg

    def goal_callback(self, msg):
        print(msg)
        self.goal = msg

File: controllers/ilqr/src/utils/ilqr.cpp
#include "ilqr/ilqr.h"

#define PI 3.1415

iLQR::iLQR(ros::NodeHandle nh, Dynamics* dynamic, Cost* cost, 
            double dt, int N)
{
    _nh = nh;
    _dynamic = dynamic;
    _cost = cost;

    _dt = dt;
    _N = N;

    _v_max = cost->get_v_max();
    _omega_max = cost->get_omega_max();

    _x    = VectorXd(_Nx);
    _x0   = VectorXd(_Nx);
    _xref = VectorXd(_Nx);

    _xs = MatrixXd(_N+1, _Nx);
    _us = MatrixXd(_N, _Nu);

    _xs_new = MatrixXd(_N+1, _Nx);
    _us_new = MatrixXd(_N, _Nu);

    _ks = MatrixXd::Zero(_N, _Nu);
    _Ks = Tensor<float, 3>(_N, _Nu, _Nx);

    _alphas = std::vector<float>(8);
    for (int i = 0; i < 8; ++i) 
        _alphas[i] = std::pow(0.5, i);

    nh.getParam("/odom/frame_id", _odom_topic);
    nh.getParam("/odom/frame_id", _frame_id);

    nh.getParam("/controller/iLQR/max_iter", _max_iter);
    nh.getParam("/controller/iLQR/tol", _tol);

    nh.getParam("/controller/iLQR/v_ref", _v_ref);

    _odom_msg.pose.pose.orientation.w = 1.0;

    _goal_msg.pose.position.x = 0.0;
    _goal_msg.pose.position.y = 0.0;
    _goal_msg.pose.orientation.w = 1.0;

    _goal_sub = _nh.subscribe("terrasentia/goal", 10, &iLQR::_goal_callback, this);
    _odom_sub = _nh.subscribe(_odom_topic, 10, &iLQR::_odom_callback, this);

    _path_pub = nh.advertise<nav_msgs::Path>("/terrasentia/path", 5);
    _cmd_vel_pub = nh.advertise<geometry_msgs::TwistStamped>("/terrasentia/cmd_vel", 5);  

}

void iLQR::run()
{
    ros::Time init = ros::Time::now();

    double odom_heading, goal_heading, heading;
    MatrixXd us(_N, _Nu);

    odom_heading = this->_get_heading(_odom_msg.pose.pose);
    goal_heading = this->_get_heading(_goal_msg.pose);

    _x0 << _odom_msg.pose.pose.position.x,
          _odom_msg.pose.pose.position.y,
          odom_heading;

    _xref << _goal_msg.pose.position.x,
             _goal_msg.pose.position.y,
             goal_heading;

    // us.setZero();

    heading = atan2(sin(odom_heading - goal_heading), cos(odom_heading - goal_heading));
    auto us0 = this->_single_shooting(heading);

    us.col(0).setConstant(_v_ref);
    us.col(1).setConstant(us0);


    this->fit(us);

    // Gen Path msg
    this->_rollout();

    geometry_msgs::PoseStamped pose;

    _path_msg.header.stamp = ros::Time::now();
    _path_msg.header.frame_id = _frame_id;


    for (int n=0; n<_N; n++)
    {
        geometry_msgs::Quaternion q_euler = tf::createQuaternionMsgFromRollPitchYaw(0.0, 0.0, _xs(n, 2));

        pose.pose.position.x = _xs(n, 0);
        pose.pose.position.y = _xs(n, 1);

        pose.pose.orientation = q_euler;

        _path_msg.poses.push_back(pose);
    }

    _path_pub.publish(_path_msg);
    _path_msg.poses.clear();

    _cmd_vel_msg.twist.linear.x = _us(0, 0);
    _cmd_vel_msg.twist.angular.z = _us(0, 1);

    _cmd_vel_pub.publish(_cmd_vel_msg);

    ROS_INFO_STREAM("Run iLQR | State " << _x0(0) << " " << _x0(1) << " " << _x0(2) << 
                            " | Reference " << _xref(0) << " " << _xref(1) << " " << _xref(2) << 
                            " | Action control " << _us(0, 0) << " " << _us(0, 1 ) << 
                            " | Time " << ros::Time::now() - init << " seconds");
}

void iLQR::fit(MatrixXd us)
{
    _us = us;

    this->_rollout();

    _J = _cost->trajectory_cost(_xs, _us, _xref);
    //ROS_INFO_STREAM("Trajectory cost: " << _J);

    for (int iter=0; iter<_max_iter; iter++)
    {

        this->_backward_pass(); 

        if (std::abs(_delta_J) < _tol)
        {
            //ROS_INFO_STREAM("Early stop");
            break;
        }

        for (float alpha : _alphas)
        {
            this->_forward_pass(alpha);
            _J_new = _cost->trajectory_cost(_xs_new, _us_new, _xref);

            //ROS_INFO_STREAM("Trying converge " << _J_new << " on " << _J << "Using regu " << regu);

            if (std::abs(_J_new) < std::abs(_J))
            {
                _J = _J_new;
                _xs = _xs_new;
                _us = _us_new;

                break;
            }
        }
    }

    ROS_INFO_STREAM("Final fit");
    // ROS_INFO_STREAM(_xs);
    ROS_INFO_STREAM(_us);
}

void iLQR::_forward_pass(float alpha)
{
    int n, i, j;
    MatrixXd Ks(_Nu, _Nx), xs_aux(1, 3);

    std::vector<float> xs(_xs.cols());
    std::vector<float> us(_us.cols());
    std::vector<DM> input;

    _xs_new.row(0) = _xs.row(0);

    for (n=0; n<_N; n++)
    {
        for (i=0; i<_Nu; i++)
            for (j=0; j<_Nx; j++)
                Ks(i, j) = _Ks(n, i, j);

        xs_aux = (_xs_new.row(n) - _xs.row(n)).transpose();

        _us_new.row(n) = _us.row(n) + (Ks * xs_aux).transpose() + alpha * _ks.row(n);

        for (i=0; i<_Nx; i++)
            xs.at(i) = _xs_new(n, i);

        for (i=0; i<_Nu; i++)
            us.at(i) = _us_new(n, i);

        input = {DM(xs), DM(us)};
        auto result = _dynamic->get_f(input);

        _xs_new.row(n+1) << result(0, 0), result(0, 1), result(0, 2);

        //ROS_INFO_STREAM("Xs: " << _xs_new.row(n+1) << "(n=" << n << ")");
        //ROS_INFO_STREAM("Us: " << _us_new.row(n) << "(n=" << n << ")");
    }
}

void iLQR::_backward_pass()
{    
    int n, i, j;
    double J=0;

    f_prime_t f_prime;
    l_prime_t l_prime;

    MatrixXd k, K;
    MatrixXd Q_x, Q_u, Q_xx, Q_uu, Q_ux; 

    std::vector<float> xs(_xs.cols());
    std::vector<float> us(_us.cols());
    std::vector<DM> input;
    
    MatrixXd Qf = _cost->get_Qf();
    
    MatrixXd s = (Qf * _xs.row(_N).transpose() - _xref).transpose();
    MatrixXd S = Qf;

    for (n=_N-1; n>=0; n--)
    {
        for (i=0; i<_Nx; i++)
            xs.at(i) = _xs(n, i) - _xref(i);

        for (i=0; i<_Nu; i++)
            us.at(i) = _us(n, i);

        input = {DM(xs), DM(us)};
        f_prime = _dynamic->get_f_prime(input);
        l_prime = _cost->get_l_prime(input);

        Q_x = l_prime.l_x + s * f_prime.f_x;
        Q_u = l_prime.l_u + s * f_prime.f_u;

        Q_x.transposeInPlace();
        Q_u.transposeInPlace();

        Q_xx = l_prime.l_xx + f_prime.f_x.transpose() * S * f_prime.f_x;
        Q_uu = l_prime.l_uu + f_prime.f_u.transpose() * S * f_prime.f_u;
        Q_ux = l_prime.l_ux + f_prime.f_u.transpose() * S * f_prime.f_x;

        k = - Q_uu.completeOrthogonalDecomposition().pseudoInverse() * Q_u;
        K = - Q_uu.completeOrthogonalDecomposition().pseudoInverse() * Q_ux;

        _ks.row(n) = k.transpose();
        for (i=0; i<_Nu; i++)
            for (j=0; j<_Nx; j++)
                _Ks(n, i, j) = K(i, j);

        s = Q_x + K.transpose() * Q_u + Q_ux.transpose() * k + K.transpose() * Q_uu * k;
        S = Q_xx + K.transpose() * Q_uu * K + K.transpose() * Q_ux + Q_ux.transpose() * K;

        s.transposeInPlace();

        J += (0.5 * (k.transpose() * Q_uu * k) + k.transpose() * Q_u)(0,0);
    }

    _delta_J = J;
    //ROS_INFO_STREAM("Final backward cost: " << J);

}

void iLQR::_rollout()
{
    int n, i;
    std::vector<float> xs(_xs.cols());
    std::vector<float> us(_us.cols());
    std::vector<DM> input;


    _xs.row(0) = _x0;

    for (n=0; n<_N; n++)
    {
        for (i=0; i<_Nx; i++)
            xs.at(i) = _xs(n, i);

        for (i=0; i<_Nu; i++)
            us.at(i) = _us(n, i);

        input = {DM(xs), DM(us)};
        auto result = _dynamic->get_f(input);
        
        _xs.row(n+1) << result(0, 0), result(0, 1), result(0, 2);
        //ROS_INFO_STREAM("Xs: " << _xs.row(n+1) << "(n=" << n << ")");
    }
}

double iLQR::_single_shooting(double err_heading)
{  
    double abs_err = abs(err_heading);
    double res = ( exp(_shoot * abs_err) - 1) * (_cost->get_omega_max() / exp(_shoot * PI));

    return (err_heading >= 0) ? -res : res;
}

double iLQR::_get_heading(geometry_msgs::Pose pose)
{    
    double roll, pitch, yaw;

    tf::Quaternion q(pose.orientation.x,
                     pose.orientation.y,
                     pose.orientation.z,
                     pose.orientation.w);

    tf::Matrix3x3 m(q);
    m.getRPY(roll, pitch, yaw);

    return yaw;
}

void iLQR::_goal_callback(const geometry_msgs::PoseStamped::ConstPtr& msg)
{
    _goal_msg = *msg;
}

void iLQR::_odom_callback(const nav_msgs::Odometry::ConstPtr& msg)
{
    _odom_msg = *msg;
}